src/platform/mbed/BLEManagerImpl.cpp

```diff
@@ -0,0 +1,1070 @@
+/*
+ *
+ *    Copyright (c) 2020-2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *          Provides an implementation of the BLEManager singleton object
+ *          for mbed platforms.
+ */
+
+#include <inttypes.h>
+#include <stdint.h>
+
+#include <platform/internal/CHIPDeviceLayerInternal.h>
+
+#if CHIP_DEVICE_CONFIG_ENABLE_CHIPOBLE
+
+#include <platform/mbed/BLEManagerImpl.h>
+
+#include <ble/CHIPBleServiceData.h>
+#include <platform/internal/BLEManager.h>
+#include <support/CodeUtils.h>
+#include <support/logging/CHIPLogging.h>
+
+// Show BLE status with LEDs
+#define _BLEMGRIMPL_USE_LEDS 0
+
+/* Undefine the BLE_ERROR_NOT_IMPLEMENTED macro provided by CHIP's
+ * src/ble/BleError.h to avoid a name conflict with Mbed-OS ble_error_t
+ * enum value. For the enum values, see:
+ * mbed-os/connectivity/FEATURE_BLE/include/ble/common/blecommon.h
+ */
+#undef BLE_ERROR_NOT_IMPLEMENTED
+// mbed-os headers
+#include "ble/BLE.h"
+#include "ble/Gap.h"
+#include "platform/Callback.h"
+#include "platform/NonCopyable.h"
+#include "platform/Span.h"
+
+#if _BLEMGRIMPL_USE_LEDS
+#include "drivers/DigitalOut.h"
+#endif
+
+using namespace ::chip;
+using namespace ::chip::Ble;
+using namespace ::chip::System;
+
+namespace chip {
+namespace DeviceLayer {
+namespace Internal {
+
+namespace {
+const UUID ShortUUID_CHIPoBLEService(0xFEAF);
+// RX = BleLayer::CHIP_BLE_CHAR_1_ID
+const UUID LongUUID_CHIPoBLEChar_RX("18EE2EF5-263D-4559-959F-4F9C429F9D11");
+const ChipBleUUID ChipUUID_CHIPoBLEChar_RX = { { 0x18, 0xEE, 0x2E, 0xF5, 0x26, 0x3D, 0x45, 0x59, 0x95, 0x9F, 0x4F, 0x9C, 0x42, 0x9F,
+                                                 0x9D, 0x11 } };
+// TX = BleLayer::CHIP_BLE_CHAR_2_ID
+const UUID LongUUID_CHIPoBLEChar_TX("18EE2EF5-263D-4559-959F-4F9C429F9D12");
+const ChipBleUUID ChipUUID_CHIPoBLEChar_TX = { { 0x18, 0xEE, 0x2E, 0xF5, 0x26, 0x3D, 0x45, 0x59, 0x95, 0x9F, 0x4F, 0x9C, 0x42, 0x9F,
+                                                 0x9D, 0x12 } };
+} // namespace
+
+#if _BLEMGRIMPL_USE_LEDS
+#include "drivers/DigitalOut.h"
+// LED1 -- toggle on every call to ble::BLE::processEvents()
+// LED2 -- on when ble::BLE::init() callback completes
+// LED3 -- on when advertising
+mbed::DigitalOut led1(LED1, 1);
+mbed::DigitalOut led2(LED2, 1);
+mbed::DigitalOut led3(LED3, 1);
+#endif
+
+struct ConnectionInfo
```

Comment:
nit: struct -> class ?

Reply:
Changed. 
I haven't found anything in the style guide about this. Could be worth adding a note, this is a pattern commonly used in Mbed OS cpp files.

Reply:
Nothing in the CHIP style guide it just struck me as odd that a struct has only private data members. What's the difference in usage between struct and class in Mbed style?

Reply:
For header files, struct are for PODs or plain interfaces class (no member and only pure virtual function) as it makes sense in both of these case to have everything public. Class is used for everything else.

For classes internal to a `cpp` file which often are _companion_ classes it makes sense to make everything public by default. 

We're fine with everything and consistency is what matter the most. So we will make sure to use `class` for everything except PODs in this codebase.